File: aemet/Aemet-Java-IvanRC/src/main/java/storage/medicion/MedicionStorageServiceCsv.java
package storage.medicion;

import config.ConfigApp;
import dto.MedicionDto;
import jdk.jshell.execution.Util;
import mapper.MedicionesMapper;
import models.Medicion;
import utils.Utils;

import java.io.*;
import java.util.ArrayList;
import java.util.List;

public class MedicionStorageServiceCsv implements MedicionStorageService {

    ConfigApp configApp = ConfigApp.getInstance();

    File file;

    String name;

    public MedicionStorageServiceCsv(String name) throws IOException {
        this.name = name;
        this.file = new File(ClassLoader.getSystemResource(name).getFile());
    }

    @Override
    public void saveAll(List<Medicion> entities) {
        // Para este caso, como no vamos a escribir en el csv, no requerimos de esta función.
    }

    @Override
    public List<Medicion> loadAll() throws FileNotFoundException {
        List<Medicion> ingredientes = new ArrayList<>();
        if(!file.exists()) return ingredientes;
        ArrayList<Integer> fecha = Utils.toFormatoFechaValido(name);
        try(BufferedReader br = new BufferedReader(new FileReader(file))){
            String line;
            while ((line = br.readLine()) != null) {
                String[] campos = line.split(";");
                ingredientes.add(
                        MedicionesMapper.toMedicion(
                                new MedicionDto(
                                        campos[0],
                                        campos[1],
                                        campos[2],
                                        campos[3],
                                        campos[4],
                                        campos[5],
                                        campos[6],
                                        fecha.get(0)+"-"+fecha.get(1)+"-"+fecha.get(2)
                                ))
                );
            }
        } catch (FileNotFoundException e) {
            throw new RuntimeException(e);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        return ingredientes;
    }
}

File: aemet/Aemet-kotlin-IvanRC/src/main/kotlin/storage/medicion/MedicionStorageServiceCsv.kt
package storage.medicion

import model.Medicion
import mu.KotlinLogging
import utils.toFormatoFechaValido
import utils.toLocalTime
import java.io.File
import java.io.IOException
import java.time.LocalDate

class MedicionStorageServiceCsv(val nombreFicheroDeResources: String): MedicionStorageService {

    val path = ClassLoader.getSystemResource(nombreFicheroDeResources).file ?: throw IOException("Fichero no encontrado.")
    private val logger = KotlinLogging.logger {}

    private lateinit var fecha: Triple<Int, Int, Int>;

    init{
        // La lista fecha, es un triple de enteros, cuyos valores son Año, Mes, Dia
        fecha = nombreFicheroDeResources.toFormatoFechaValido()
    }

    override fun saveAll(entities: List<Medicion>) {
        // Para este caso, como no vamos a escribir en el csv, no requerimos de esta función.
        TODO("Not yet implemented")
    }

    override fun loadAll(): List<Medicion> {
        logger.debug { "Se cargan todas las mediciones del archivo $nombreFicheroDeResources" }
        //El csv no tiene cabecera
        val file = File(path)
        if(!file.exists()) return emptyList()
        return file.readLines()
            .map { it.split(";").map { it.trim() } }
            .map { campos ->
                Medicion(
                    poblacion = campos[0],
                    provincia = campos[1],
                    temperaturaMax = campos[2].toDouble(),
                    // Es la función que tengo en mi Utils
                    horaTempMax = campos[3].toLocalTime(),
                    temperaturaMin = campos[4].toDouble(),
                    horaTempMin = campos[5].toLocalTime(),
                    precipitacion = campos[6].toDouble(),
                    dia = LocalDate.of(fecha.first, fecha.second, fecha.third)
                )
            }
    }
}